[SPARK-10983] Unified memory manager

andrewor14 opened the pull request:
This patch unifies the memory management of the storage and execution regions such that either side can borrow memory from each other. When memory pressure arises, storage will be evicted in favor of execution. To avoid regressions in cases where storage is crucial, we dynamically allocate a fraction of space for storage that execution cannot evict. Several configurations are introduced:
- **spark.memory.fraction (default 0.75)**: ​fraction of the heap space used for execution and storage. The lower this is, the more frequently spills and cached data eviction occur. The purpose of this config is to set aside memory for internal metadata, user data structures, and imprecise size estimation in the case of sparse, unusually large records.
- **spark.memory.storageFraction (default 0.5)**: size of the storage region within the space set aside by `s​park.memory.fraction`. ​Cached data may only be evicted if total storage exceeds this region.
- **spark.memory.useLegacyMode (default false)**: whether to use the memory management that existed in Spark 1.5 and before. This is mainly for backward compatibility.

For a detailed description of the design, see [SPARK-10000](https://issues.apache.org/jira/browse/SPARK-10000). This patch builds on top of the `MemoryManager` interface introduced in #9000.

Commit:
Add UnifiedMemoryManager skeleton

Commit:
First implementation of UnifiedMemoryManager

As of this commit, acquiring execution memory may cause cached
blocks to be evicted. There are still a couple of TODOs, notably
figuring out what the `maxExecutionMemory` and `maxStorageMemory`
should actually be. They may not be fixed since either side can
now borrow from the other.

Commit:
Use a global lock for all memory manager interactions

Without this, we cannot both avoid deadlocks and race conditions,
because each individual component ShuffleMemoryManager,
MemoryStore and MemoryManager all have their own respective
locks.

This commit allows us to simplify several unintuitive control
flows that were introduced to avoid acquiring locks in different
orders. Since we have only one lock now, these code blocks can
be significantly simplified.

A forseeable downside to this is parallelism is reduced,
but memory acquisitions and releases don't actually occur that
frequently, so performance should not suffer noticeably.
Additional investigations about this should ensue.

Commit:
Use a dynamic max for execution and storage

Previously, ShuffleMemoryManager will allow each task to acquire
up to 1/N of the entire storage + execution region. What we want
is more like 1/N of the space not occupied by storage, since the
"max" now varies over time.

Commit:
Fix notifyAll() IllegalMonitorStateException

This happened because we were still calling `this.notifyAll()`
in ShuffleMemoryManager when we were holding the `memoryManager`
lock. Easy fix.

Commit:
Minor: update a few comments

Commit:
Register blocks evicted by execution

Commit:
Minor: more comment updates

Commit:
Add tests for UnifiedMemoryManager + TODOs

Tests are passing in this commit, but there are follow-ups that
need to be done (see TODOs added in this commit). More tests will
be added in the future.

Commit:
Clean up test code, resolve TODOs

As of this commit all *MemoryManagerSuite's are documented and
pass tests.

Commit:
Fix tests

TaskContext was not stubbed correctly in ShuffleMemoryManagerSuite.
In particular, this patch added some code that does some things
to the active TaskMetrics, but this was not part of the mocked
TaskContext object.

Commit:
Minor: Add TODOs, fix a few comments

Commit:
Merge branch 'master' of github.com:apache/spark into unified-memory-manager

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43602 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43602/consoleFull) for   PR 9084 at commit [`01ff533`](https://github.com/apache/spark/commit/01ff5335b2b6554ed9388b015145b8dfd9bc1977).

Commit:
Documentation + enable by default

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Log deprecation warning for old memory fraction configs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43607 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43607/consoleFull) for   PR 9084 at commit [`059ae0d`](https://github.com/apache/spark/commit/059ae0da40e981306a060bc9f7dafd7639263cf0).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43606/
Test FAILed.

SparkQA commented:
[Test build #43602 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43602/console) for   PR 9084 at commit [`01ff533`](https://github.com/apache/spark/commit/01ff5335b2b6554ed9388b015145b8dfd9bc1977).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43602/
Test PASSed.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -509,6 +512,11 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
+  // Deprecation message for memory fraction configs used in the old memory management model
+  private val deprecatedMemoryFractionMessage =
+    "As of Spark 1.6, execution and storage memory management are unified. " +
+      "All memory fractions used in the old model are now deprecated and no longer read."
```
The old configurations will still be respected in legacy mode, so this is slightly ambiguous / confusing. Is there an easy way to avoid the warning if the legacy mode configuration is turned on? If not, I suppose we could just expand the deprecation message to mention this corner case, perhaps by just appending a "(unless spark.XX.YY is enabled)" at the end.

andrewor14 replied:
I can print a different warning if legacy mode is on

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:
```diff
@@ -40,19 +41,38 @@ private[spark] abstract class MemoryManager {
+   */
+  def maxStorageMemory: Long
+
+  // TODO: avoid passing evicted blocks around to simplify method signatures (SPARK-10985)
```
Good call on deferring this to a followup task.

SparkQA commented:
[Test build #43607 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43607/console) for   PR 9084 at commit [`059ae0d`](https://github.com/apache/spark/commit/059ae0da40e981306a060bc9f7dafd7639263cf0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43607/
Test FAILed.

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:
```diff
@@ -72,46 +92,62 @@ private[spark] abstract class MemoryManager {
   def acquireUnrollMemory(
```
Given that `acquireUnrollMemory` appears to act as a synonym for `acquireStorageMemory` in the current implementation, it might be worth adding a brief comment above this method to explain that this extra method exists in order to give us the future flexibility to account for unroll memory differently in the future.

andrewor14 replied:
actually, it's more than a synonym. In `StaticMemoryManager` it's required to preserve existing behavior where unrolling doesn't evict all the blocks.

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:
```diff
@@ -72,46 +92,62 @@ private[spark] abstract class MemoryManager {
    * Release all storage memory acquired.
    */
-  def releaseStorageMemory(): Unit
+  def releaseStorageMemory(): Unit = synchronized {
```
Minor nit (which I guess existed in the old code), but what do you think about naming this `releaseAllStorageMemory`?

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:
```diff
@@ -0,0 +1,138 @@
+ * Similarly, execution can borrow as much storage memory as is free. However, execution
+ * memory is *never* evicted by storage due to the complexities involved in implementing this.
+ * The implication is that attempts to cache blocks may fail if execution has already eaten
+ * up most of the storage space, in which case the new blocks will be evicted directly.
```
"evicted immediately"? I also wonder if this should clarify that the new blocks will not be cached if their storage level is MEMORY_ONLY and will be spilled to disk if their storage level permits that.

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:
```diff
@@ -0,0 +1,138 @@
+    val shouldEvictStorage = totalFreeMemory < numBytes && memoryBorrowedByStorage > 0
+    if (shouldEvictStorage) {
+      val spaceToEnsure = math.min(numBytes, memoryBorrowedByStorage)
+      memoryStore.ensureFreeSpace(spaceToEnsure, evictedBlocks)
```
Is it possible that this call could be expensive? Just wondering whether it's safe to be performing inside of the `synchronized` block. Note that we can always optimize the locking later, too.

JoshRosen replied:
Also, `ensureFreeSpace` has a return value. Should we check it?

JoshRosen replied:
Another possible concern: I wonder whether it's possible for there to be a deadlock between `MemoryManager.this` and `MemoryStore.accountingLock`. For instance, `MemoryStore. reserveUnrollMemoryForThisTask` first synchronizes on the accounting lock and then calls `MemoryManager.acquireUnrollMemory`, while at the same time `MemoryManager.acquireExecutionMemory` calls `MemoryStore.ensureFreeSpace`.

andrewor14 replied:
there's no more accounting lock :)

JoshRosen replied:
Oh, duh. That's what I get for looking at the unified diff.

Looks safe to me now.

andrewor14 replied:
as for whether it's potentially expensive, yes it may be. However, note that you only do this under memory pressure, meaning your tasks would have otherwise spilled anyway, so in either case we'll end up doing some I/O

andrewor14 replied:
(for the `ensureFreeSpace` return value, I looked into it a little more but I don't think there's anything useful we can do with it. The problem is that in unrolling it could return `false` but still end up giving you enough space.)

JoshRosen commented on core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:
```diff
@@ -0,0 +1,138 @@
+ * The region shared between execution and storage is a fraction of the total heap space
+ * configurable through `spark.memory.fraction` (default 0.75). The position of the boundary
+ * within this space is further determined by `spark.memory.storageFraction` (default 0.5).
+ * This means the size of the storage region is 0.75 * 0.5 = 0.375 of the heap space by default.
```
This might be a really cool place to put a ASCII-art diagram illustrating the breakdown (like a horizonal stacked barchart). I can help make one.

andrewor14 replied:
thanks :+1:

JoshRosen commented on core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala:
```diff
@@ -65,7 +66,7 @@ class ShuffleMemoryManager protected (
    * happen if the number of tasks increases but an older task had a lot of memory already.
    */
-  def tryToAcquire(numBytes: Long): Long = synchronized {
+  def tryToAcquire(numBytes: Long): Long = memoryManager.synchronized {
```
This slightly-confusing synchronization on `memoryManager` isn't a huge deal since it's going to be removed by the memory manager unification patch.

JoshRosen commented on core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala:
```diff
@@ -73,15 +74,18 @@ class ShuffleMemoryManager protected (
     // Keep looping until we're either sure that we don't want to grant this request (because this
     // task would have more than 1 / numActiveTasks of the memory) or we have enough free
     // memory to give it (we always let each task get at least 1 / (2 * numActiveTasks)).
+    // TODO: simplify this to limit each task to its own slot
```
Can you elaborate on what changes this might require? Will we have to effectively port the current cross-task accounting logic into the StaticMemoryManager in order to preserve the existing behavior for legacy mode?

andrewor14 replied:
no, this is largely orthogonal to the memory management task. Limiting a task to its slot affects both legacy and unified mode. The motivation for doing this is to simplify the logic here, so we shouldn't bend ourselves backwards to try to exactly preserve the legacy behavior.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -37,15 +37,14 @@ private case class MemoryEntry(value: Any, size: Long, deserialized: Boolean)
 
   // A mapping from taskAttemptId to amount of memory used for unrolling a block (in bytes)
-  // All accesses of this map are assumed to have manually synchronized on `accountingLock`
+  // All accesses of this map are assumed to have manually synchronized on `memoryManager`
```
Does this imply that `unrollMemoryMap` should live inside of `memoryManager`?

andrewor14 replied:
maybe it should, but I think it's largely irrelevant in the future when we have datasets. I would prefer to keep it in `MemoryStore` for now to minimize the changes. (we can always move it later)

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -365,16 +372,13 @@ private[spark] class MemoryStore(blockManager: BlockManager, memoryManager: Memo
-      // have to revisit this assumption. See SPARK-10983 for more context.
+      // The synchronization on `memoryManager` here guarantees that the release and acquire
+      // happen atomically. This relies on the assumption that all memory acquisitions are
+      // synchronized on the same lock.
```
It's nice that this simplification resulted from the single-lock approach. I'm happy to stick with that simple approach until perf. data suggests that we should do something more complex.

JoshRosen commented:
It looks like the test failures are occurring in the external sorter suites. Some of these tests exercise spilling-related logic (such as ensuring that spill files are cleaned up); in order to induce spilling, these tests configured the legacy memory fractions to be really small. In order to fix these tests, we can either enable legacy mode for those tests only or we can modify the tests to set the new configurations. Ideally these sorter unit tests wouldn't be relying on memory management behavior in order to test spill cleanup, but they were written a long time ago. We could refactor these tests, but I'd prefer a band-aid solution for now.

Commit:
Fix tests

Many of these tests relied on setting a very low shuffle memory
fraction. Since that config is now deprecated and not read, the
tests now fail because things aren't spilling. This approach is,
however, quite brittle as it depends on the heap size of the test
JVM. Instead, we should explicitly set a limit on the memory used.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43622/consoleFull) for   PR 9084 at commit [`24a391c`](https://github.com/apache/spark/commit/24a391cbba36812d555d32153daa5c132b3e608d).

Commit:
Remove more usages of old deprecated configs

Unfortunately not all affected tests can be easily rewritten.
Instead, this commit adds TODO's in places where the tests may
not actually do anything so we can fix them in the future
(SPARK-11078).

Commit:
Warn against deprecated configs only in non-legacy mode

Commit:
Address comments

Commit:
Merge branch 'master' of github.com:apache/spark into unified-memory-manager

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43631/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

SparkQA commented:
[Test build #43622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43622/console) for   PR 9084 at commit [`24a391c`](https://github.com/apache/spark/commit/24a391cbba36812d555d32153daa5c132b3e608d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43622/
Test FAILed.

SparkQA commented:
[Test build #43631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43631/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43631/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #1889 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1889/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

SparkQA commented:
[Test build #1888 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1888/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

SparkQA commented:
[Test build #1891 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1891/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

SparkQA commented:
[Test build #1890 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1890/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43646/consoleFull) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).

SparkQA commented:
[Test build #43646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43646/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43646/
Test FAILed.

SparkQA commented:
[Test build #1890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1890/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1888/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1889/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1891 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1891/console) for   PR 9084 at commit [`face129`](https://github.com/apache/spark/commit/face129d898db2ebcea200ed448ec2e78c288c73).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

JoshRosen commented:
LGTM. I'm going to merge this now. There's some followup tasks to do in terms of refactoring some old spilling tests, but there's followup tasks for those and we'll do them soon.

pzz2011 commented on core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:
```diff
@@ -40,6 +41,10 @@ private[spark] abstract class MemoryManager {
   }
 
+  // Amount of execution/storage memory in use, accesses must be synchronized on `this`
+  protected var _executionMemoryUsed: Long = 0
```
why here is protected?

srowen replied:
@pzz2011 you're making several comments on old PRs. Generally people won't see that and it's not the place for discussion anyway. If you can formulate a specific question beyond "why is the code this way?" ask on user@.
